src/esthri/src/lib.rs

```diff
@@ -1084,54 +1094,104 @@ where
+                break;
+            };
+            if stat.is_dir() {
+                continue;
+            }
+            if entry.path_is_symlink() {
+                warn!("symlinks are ignored");
+                continue;
+            }
+            let path = format!("{}", entry.path().display());
+            debug!("local path={}", path);
+            if process_globs(&path, glob_includes, glob_excludes).is_some() {
+                yield Ok((entry.path().display().to_string(), ListingMetadata::none()));
+            }
+        }
+    }
+}
+
+fn map_paths_to_etags<StreamT>(
+    input_stream: StreamT,
+) -> impl Stream<Item = JoinHandle<MapEtagResult>>
 where
-    T: S3 + Send + Clone,
+    StreamT: Stream<Item = Result<(String, Option<ListingMetadata>)>>,
 {
-    let directory = directory.as_ref();
-    if !key.ends_with(FORWARD_SLASH) {
-        return Err(Error::DirlikePrefixRequired);
-    }
-    for entry in WalkDir::new(directory) {
-        let entry = entry?;
-        let stat = entry.metadata()?;
-        if stat.is_dir() {
-            continue;
-        }
-        // TODO: abort if symlink?
-        let path = format!("{}", entry.path().display());
-        debug!("local path={}", path);
-        let path = process_globs(&path, glob_includes, glob_excludes);
-        if let Some(path) = path {
-            let remote_path = Path::new(key);
-            let stripped_path = entry.path().strip_prefix(&directory);
+    input_stream.map(|params| {
+        task::spawn_blocking(move || {
+            let (path, metadata) = params?;
+            let local_etag = s3_compute_etag(&path);
+            Ok((path, local_etag, metadata))
+        })
+    })
+}
+
+fn local_to_remote_sync_tasks<ClientT, StreamT>(
+    s3: ClientT,
+    bucket: String,
+    key: String,
+    directory: PathBuf,
+    dirent_stream: StreamT,
+) -> impl Stream<Item = impl Future<Output = Result<()>>>
+where
+    ClientT: S3 + Send + Clone,
+    StreamT: Stream<Item = StdResult<MapEtagResult, JoinError>>,
+{
+    dirent_stream
+        .map(move |entry| {
+            (
+                s3.clone(),
+                bucket.clone(),
+                key.clone(),
+                directory.clone(),
+                entry.unwrap(),
+            )
+        })
+        .map(|clones| async move {
+            let (s3, bucket, key, directory, entry) = clones;
+            let (path, local_etag, _metadata) = entry?;
+            let path = Path::new(&path);
+            let remote_path = Path::new(&key);
+            let stripped_path = path.strip_prefix(&directory);
             let stripped_path = match stripped_path {
                 Err(e) => {
                     warn!("unexpected: failed to strip prefix: {}", e);
-                    continue;
+                    return Ok(());
                 }
                 Ok(result) => result,
             };
             let stripped_path = format!("{}", stripped_path.display());
```

Comment:
I know this was not in your PR but I stumbled on a neat way for converting a Path/PathBuf to a Cow using ".to_lossy_string()" and borrowing this gives you an &str.

Reply:
Thanks, cleaned up what I could find